### ISSUE TITLE
C# refactoring project

### ISSUE THREAD

**Issue by cadatoiva** [Jackett/Jackett#6329]
Not an issue, more of a question. I want to help contribute to the C# team and have forked the repo. First thing I want to do is some non-critical refactoring to make the codebase more consistent. I have ReSharper to help with this, and was wondering if you all have any tips, requirements, or style guides to help with my efforts.

Some basic things I've already put some time into are the following non-critical changes:

- Cleanup whitespace and indenting
- Change to var for local variable declaration (to match C# standard style guide)
- Remove unnecessary parenthesis, commas, semicolons, and using statements.

**Comment by garfield69** [comment]
@kaso17  Please take some time off your busy schedule to read this ticket and answer as you see fit.
As you know I am not a c# dev and have no opinion one way or the other about this topic.
Thanks.

**Comment by ngosang** [comment]
@cadatoiva thank you for the interest. This project needs c# developers. Before doing big changes, what about fixing some open issues related to c# to prove yourself?

**Comment by ngosang** [comment]
Related => https://github.com/Jackett/Jackett/pull/6816#issuecomment-570907542
@junglebus what do you think?

**Comment by junglebus** [comment]
@cadatoiva sounds great what you are suggesting
I'm only a hobby tinkerer, so you sound much more knowledgeable 
At the moment there is absolutely no style guide or convention. Everyone just adds and changes in their own style
Being an open source program with many contributors what do you think about going for a rules based approach to make it easier for new contributors. That way it can be included in the code and everyone gets it. Whereas if you just cleanup using your Resharper (which no elese probably has a license for) you'll be cleaning up everyone elses messes forever. I'm pretty sure you can even get to the the point where PR's need to pass the ruleset to be merged. I think the Jellyfin guys have put a lot of work into this https://github.com/jellyfin/jellyfin/blob/master/jellyfin.ruleset But not sure if thats a good starting point or something else?


A little bit off topic but slightly related. I'd like to be more confident when I change stuff that I haven't broken anything and hence would just like to have one simple integration test that runs after a PR, The integration test would just install Jackett and then use Selenium to open the manual search page and search a safe linux distro tracker. It would run for windows, linux, macOs and Mono
@garfield69 would you be open to switching the build system from Appveyor to Azure pipelines if I got it going on my fork first. Could even run them in parallel for a while. Been having a look at Jellyfin and Radarr (who I think are technically the closest to Jackett) and they are doing some really cool stuff with Azure pipelines. Azure has every version of Mono since 4.8 installed https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/macos/macos-10.14-Readme.md#mono e.g. Radarr test against four versions of Mono https://github.com/Radarr/Radarr/blob/aphrodite/azure-pipelines.yml#L528

**Comment by cadatoiva** [comment]
I agree with deciding on using a rules based style guide and adding wiki/readme entry detailing what is in those rules. It's coming to that initial consensus as to what the style guide should be for the project that is the hardest part. ReSharper and VS Community and other tools can all accept custom style rules and apply them across the project before submission. There's a style bot on GH that can filter PRs that don't follow the project style guide as well.

I like most of what I saw in this style guide repo https://github.com/MakingSense/code-style-guides/tree/master/CSharp (but I'm still partial to tabs over spaces, which is apparently against the MS sanctioned style guide, which says save files to repo as spaces). This nu-get package would automatically make any contributor bound to whatever style guide we decide on, so the code won't compile without being in the correct style https://github.com/DotNetAnalyzers/StyleCopAnalyzers

Once we make a decision and integrate a ruleset, I can import that into ReSharper and run the solution wide code cleanup command to make the whole project fit the code style in one go for the first run, after that the nu-get StyleCop + rule set will enforce it for all future code.

**Comment by junglebus** [comment]
@cadatoiva MakingSense one looked good to me and not too onerous to begin with. Even if its not perfect, will be better than what we currently have and will be good to have some consistency
If the initial change is too massive, maybe enable a couple of rules at a time???

**Comment by cadatoiva** [comment]
I'm going to do some testing on my local fork and see how it plays out.

**Comment by garfield69** [comment]
>would you be open to switching the build system from Appveyor to Azure pipelines if I got it going on my fork first

yes. 
I personally have no interest in what build method is used to generate the packages, as long as it work ;-)

**Comment by cadatoiva** [comment]
So, after searching through some documentation, I've pretty much narrowed down how to add a project wide implementation of a style guide and StyleCop implementation. If we add the style guide then it will only affect changes done in the future until we go through with code cleanup, format code or `dotnet format`

I can make a PR that implements the style guide into the solution and updates a document outlining what the chosen guide is. That PR won't change any of the project code yet, until we agree on the style and merge it. Then I'll submit another PR that is just the result of running `dotnet format` when we've merged the selected style.

StyleCop will issue compile warnings and errors based on whatever conventions we choose for those. It involves adding a nu-get package and the style, and probably won't be able to be integrated without making the required changes, so that will be done separately from the non-blocking style guide.

**Comment by junglebus** [comment]
@cadatoiva sounds good, nice work

**Comment by cadatoiva** [comment]
Just an update guys. I am still working on this, it's just been a real chore. I got the .editorconfig up and working on the 14th. The real problem is twofold: I'm a noob at markdown and the CONTRIBUTING.md is kicking my butt.

I'm almost done at this point, but GitHub said that a contributing file would create a "please read this first" banner on issues and pull requests, so I've been looking at the contributing files at 10 fairly large projects for inspiration on what to do.

As a result the PR for the style guide will move some things out of ISSUE_TEMPLATE and README into CONTRIBUTING. I should have the whole thing ready for a PR + review in the next day or two.

**Comment by cadatoiva** [comment]
List of all Pull Requests tied to this issue for reference on conversations made on individual PRs

#7045 Implemented style guidelines
#7122 Failed attempt to format all at once
#7145 run dotnet format on solution
#7159 sort and remove usings

**Comment by Lonli-Lokli** [comment]
Have you implemented styleguide integration, ie breaking the build when new changes do not follow required styles?

**Comment by ngosang** [comment]
Not yet because the current code base don't follow the styleguide. We are doing progress but this is not a priority, we have a lot of more important issues open. Any help is welcome.

**Comment by Lonli-Lokli** [comment]
@ngosang is current styleguide required for all parts of application?

**Comment by ngosang** [comment]
Yes
